keps/sig-network/20191104-iptables-no-cluster-cidr.md

```diff
@@ -0,0 +1,282 @@
+
+This enhancement proposes ways to achieve similar goals without tracking the pod cluster CIDR to do so.
+
+## Motivation
+
+The idea that makes kubernetes networking model unique and powerful is the concept of each pod having its own IP, with all the pod IPs being natively routable within the cluster. The service chains in iptable rules depend on this capability by assuming that they can treat all the endpoints of a cluster as being equivalent and load balance service traffic across all the endpoints, but just translating destination to the pod IP address.
+
+While this is powerful, it also means pod IP addresses are in many cases the constraining resource for cluster creation and scale. It would be valuable for implementations to have different strategies for managing pod IP addresses that can adapt to different environment needs.
+
+Some examples of use cases:
+
+   * Creating a cluster out of many disjoint ranges instead of a single range.
+   * Expanding a cluster with more disjoint ranges after initial creation.
+
+Not having to depend on the cluster pod CIDR for routing service traffic would effectively de-couple pod IP management and allocation strategies from service management and routing. 
+Which in turn would mean that it would be far cheaper to evolve the IP allocation schemes while
+sharing the same service implementation, thus significantly lowering the bar for adoption of
+alternate schemes.
+
+Alternate implementations that don’t use iptables could also adopt this same reasoning to not have to track the cluster CIDR for routing cluster traffic.
+
+### Goals
+
+   * Not having to depend on the cluster pod CIDR for iptable rules and cluster traffic routing.
+
+### Non-Goals
+
+   * Providing alternate models of IP allocation schemes for pod CIDR.
+   * Enhancing current allocators to handle disjoint ranges.
+   * Enhancing current allocators to add additional ranges after cluster creation.
+   * Changing current assumptions around having a single pod CIDR per node.
+
+## Proposal
+
+As stated above, the goal is to re-implement the functionality called out in the summary, but in a 
+way that does not depend on a pod cluster CIDR. The essence of the proposal is that for the 
+first two cases in iptables implementation and first case in ipvs, we can replace the `-s proxier.clusterCIDR` with some notion of node local pod traffic.
+
+The core logic in these cases is “how to determine” cluster originated traffic from non-cluster originated ones. The proposal is that tracking pod traffic generated from within the node is sufficient to determine cluster originated traffic.
+For the first two use cases in iptables and first use case in ipvs, we provide alternatives to using the proxier.clusterCIDR in one of the following ways to determine cluster originated traffic
+
+   1. `-s node.podCIDR` (where node podCIDR is used for allocating pod IPs within the node)
+   2. `--in-interface prefix+` (where all pod interfaces start with same prefix)
```

Comment:
this topic was discussed in this issue and I've created a prototype to try to find the prefix of the pods interfaces https://github.com/kubernetes/kubernetes/issues/79866#issuecomment-522599553
My doubt is if its possible to "force" a unique interface prefix for the pods interfaces :thinking:

Reply:
I believe it is possible. By default p2p sets up a random name at https://github.com/containernetworking/plugins/blob/v0.8.3/plugins/main/ptp/ptp.go#L69 by calling SetupVeth. But there is an alternate function called SetupVethWithName at https://github.com/containernetworking/plugins/blob/v0.8.3/pkg/ip/link_linux.go#L133 which can be used to set it with a name, which then be used to have a common prefix + random string or some logic thereof to achieve this.

Reply:
@cadmuxe

Reply:
I suggest allowing multiple prefixes in a list here.  We've found that useful in Calico where we also use the interface prefix to identify pod traffic.  For example, if the cluster is being migrated from one CNI to another and some nodes are using old prefix, some new.

Reply:
good point. yes having multiple prefixes makes sense to ease migration like use cases. Will update the KEP.

Reply:
updated.

Reply:
having updated the flags to take list of CIDRs or prefixes, marking as resolved.